cpp/src/arrow/compute/kernels/vector_sort.cc

```diff
@@ -1778,6 +1799,711 @@ class SortIndicesMetaFunction : public MetaFunction {
+    }
+    switch (args[0].kind()) {
+      case Datum::ARRAY: {
+        if (select_k_options.is_top_k()) {
+          return SelectKth<SortOrder::Descending>(*args[0].make_array(), select_k_options,
+                                                  ctx);
+        } else {
+          return SelectKth<SortOrder::Ascending>(*args[0].make_array(), select_k_options,
+                                                 ctx);
+        }
+      } break;
+      case Datum::CHUNKED_ARRAY: {
+        if (select_k_options.is_top_k()) {
+          return SelectKth<SortOrder::Descending>(*args[0].chunked_array(),
+                                                  select_k_options, ctx);
+        } else {
+          return SelectKth<SortOrder::Ascending>(*args[0].chunked_array(),
+                                                 select_k_options, ctx);
+        }
+      } break;
+      case Datum::RECORD_BATCH:
+        return SelectKth(*args[0].record_batch(), select_k_options, ctx);
+        break;
+      case Datum::TABLE:
+        return SelectKth(*args[0].table(), select_k_options, ctx);
+        break;
+      default:
+        break;
+    }
+    return Status::NotImplemented(
+        "Unsupported types for sort_indices operation: "
+        "values=",
+        args[0].ToString());
+  }
+
+ private:
+  template <SortOrder sort_order>
+  Result<Datum> SelectKth(const Array& array, const SelectKOptions& options,
+                          ExecContext* ctx) const {
+    Datum output;
+    ArraySelecter<sort_order> selecter(ctx, array, options, &output);
+    ARROW_RETURN_NOT_OK(selecter.Run());
+    return output;
+  }
+
+  template <SortOrder sort_order>
+  Result<Datum> SelectKth(const ChunkedArray& chunked_array,
+                          const SelectKOptions& options, ExecContext* ctx) const {
+    Datum output;
+    ChunkedArraySelecter<sort_order> selecter(ctx, chunked_array, options, &output);
+    ARROW_RETURN_NOT_OK(selecter.Run());
+    return output;
+  }
+  Result<Datum> SelectKth(const RecordBatch& record_batch, const SelectKOptions& options,
+                          ExecContext* ctx) const {
+    Datum output;
+    RecordBatchSelecter selecter(ctx, record_batch, options, &output);
+    ARROW_RETURN_NOT_OK(selecter.Run());
+    return output;
+  }
+  Result<Datum> SelectKth(const Table& table, const SelectKOptions& options,
+                          ExecContext* ctx) const {
+    Datum output;
+    TableSelecter selecter(ctx, table, options, &output);
+    ARROW_RETURN_NOT_OK(selecter.Run());
+    return output;
+  }
+};
+
+class SelectKMetaFunction : public MetaFunction {
+ public:
+  SelectKMetaFunction()
+      : MetaFunction("select_k", Arity::Unary(), &select_k_doc, &kDefaultSelectKOptions) {
+  }
+
+  Result<Datum> ExecuteImpl(const std::vector<Datum>& args,
+                            const FunctionOptions* options,
+                            ExecContext* ctx) const override {
+    SelectKthMetaFunction impl;
```

Comment:
Can we just inline the above class here? It doesn't seem useful to split them. (Once we remove TopKMetaFunction et al)